File: JAVA/lld-examples/hard/distributed-chat-system/src/main/java/com/systemdesign/lld/chat/model/Message.java
package com.systemdesign.lld.chat.model;

import java.time.LocalDateTime;
import java.util.*;

/**
 * Message model representing a chat message in the distributed system
 */
public class Message {
    public enum MessageType {
        PRIVATE_MESSAGE,
        ROOM_MESSAGE,
        SYSTEM_MESSAGE,
        NOTIFICATION
    }
    
    public enum MessageStatus {
        PENDING,
        SENT,
        DELIVERED,
        READ,
        FAILED
    }
    
    private final String messageId;
    private final User sender;
    private String content;
    private final MessageType messageType;
    private String recipientId;
    private String roomId;
    private final LocalDateTime timestamp;
    private MessageStatus status;
    private LocalDateTime editedAt;
    private boolean isDeleted;
    
    // Message metadata
    private String serverId;
    private int deliveryAttempts;
    private final int maxDeliveryAttempts = 3;
    
    // Reply information
    private String replyToMessageId;
    private String threadId;
    
    // Media attachments
    private List<String> attachments;
    
    // Reactions
    private Map<String, String> reactions; // userId -> emoji
    
    // Recipient reference (for private messages)
    private User recipient;
    
    public Message(User sender, String content, MessageType messageType) {
        this.messageId = UUID.randomUUID().toString();
        this.sender = sender;
        this.content = content;
        this.messageType = messageType;
        this.timestamp = LocalDateTime.now();
        this.status = MessageStatus.PENDING;
        this.isDeleted = false;
        this.deliveryAttempts = 0;
        this.attachments = new ArrayList<>();
        this.reactions = new HashMap<>();
    }
    
    /**
     * Set message recipient for private messages
     */
    public void setRecipient(User recipient) {
        this.recipient = recipient;
        this.recipientId = recipient.getUserId();
    }
    
    /**
     * Set room ID for room messages
     */
    public void setRoomId(String roomId) {
        this.roomId = roomId;
    }
    
    /**
     * Mark message as sent
     */
    public void markAsSent(String serverId) {
        this.status = MessageStatus.SENT;
        this.serverId = serverId;
    }
    
    /**
     * Mark message as delivered
     */
    public void markAsDelivered() {
        this.status = MessageStatus.DELIVERED;
    }
    
    /**
     * Mark message as read
     */
    public void markAsRead() {
        this.status = MessageStatus.READ;
    }
    
    /**
     * Mark message as failed
     */
    public void markAsFailed() {
        this.status = MessageStatus.FAILED;
    }
    
    /**
     * Increment delivery attempts
     */
    public void incrementDeliveryAttempts() {
        this.deliveryAttempts++;
    }
    
    /**
     * Check if message can be retried for delivery
     */
    public boolean canRetryDelivery() {
        return status == MessageStatus.PENDING && deliveryAttempts < maxDeliveryAttempts;
    }
    
    /**
     * Edit message content
     */
    public void editContent(String newContent) {
        this.content = newContent;
        this.editedAt = LocalDateTime.now();
    }
    
    /**
     * Delete message
     */
    public void deleteMessage() {
        this.isDeleted = true;
        this.content = "[Message deleted]";
    }
    
    /**
     * Add reaction to message
     */
    public void addReaction(String userId, String emoji) {
        reactions.put(userId, emoji);
    }
    
    /**
     * Remove reaction from message
     */
    public void removeReaction(String userId) {
        reactions.remove(userId);
    }
    
    /**
     * Get reaction count
     */
    public int getReactionCount() {
        return reactions.size();
    }
    
    /**
     * Add attachment to message
     */
    public void addAttachment(String attachment) {
        attachments.add(attachment);
    }
    
    /**
     * Check if message is a private message
     */
    public boolean isPrivateMessage() {
        return messageType == MessageType.PRIVATE_MESSAGE;
    }
    
    /**
     * Check if message is a room message
     */
    public boolean isRoomMessage() {
        return messageType == MessageType.ROOM_MESSAGE;
    }
    
    /**
     * Check if message is a system message
     */
    public boolean isSystemMessage() {
        return messageType == MessageType.SYSTEM_MESSAGE;
    }
    
    /**
     * Get display content (truncated if too long)
     */
    public String getDisplayContent() {
        if (isDeleted) {
            return "[Message deleted]";
        }
        
        int maxLength = 100;
        if (content.length() > maxLength) {
            return content.substring(0, maxLength) + "...";
        }
        return content;
    }
    
    // Getters
    public String getMessageId() { return messageId; }
    public User getSender() { return sender; }
    public String getContent() { return content; }
    public MessageType getMessageType() { return messageType; }
    public String getRecipientId() { return recipientId; }
    public String getRoomId() { return roomId; }
    public LocalDateTime getTimestamp() { return timestamp; }
    public MessageStatus getStatus() { return status; }
    public LocalDateTime getEditedAt() { return editedAt; }
    public boolean isDeleted() { return isDeleted; }
    public String getServerId() { return serverId; }
    public int getDeliveryAttempts() { return deliveryAttempts; }
    public String getReplyToMessageId() { return replyToMessageId; }
    public String getThreadId() { return threadId; }
    public List<String> getAttachments() { return new ArrayList<>(attachments); }
    public Map<String, String> getReactions() { return new HashMap<>(reactions); }
    public User getRecipient() { return recipient; }
    
    // Setters
    public void setReplyToMessageId(String replyToMessageId) { this.replyToMessageId = replyToMessageId; }
    public void setThreadId(String threadId) { this.threadId = threadId; }
    
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Message message = (Message) o;
        return Objects.equals(messageId, message.messageId);
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(messageId);
    }
    
    @Override
    public String toString() {
        return "Message{" +
                "messageId='" + messageId + '\'' +
                ", sender=" + sender.getDisplayName() +
                ", type=" + messageType +
                ", content='" + getDisplayContent() + '\'' +
                '}';
    }
} 

File: Python/lld-examples/hard/distributed-chat-system/message.py
"""
Message class and related enums for the distributed chat system
"""

import uuid
from datetime import datetime
from enum import Enum
from typing import Optional
from user import User


class MessageType(Enum):
    """Enum for message types"""
    PRIVATE_MESSAGE = "private_message"
    ROOM_MESSAGE = "room_message"
    SYSTEM_MESSAGE = "system_message"
    NOTIFICATION = "notification"


class MessageStatus(Enum):
    """Enum for message delivery status"""
    PENDING = "pending"
    SENT = "sent"
    DELIVERED = "delivered"
    READ = "read"
    FAILED = "failed"


class Message:
    """Message class representing a chat message"""
    
    def __init__(self, sender: User, content: str, message_type: MessageType, 
                 recipient_id: Optional[str] = None, room_id: Optional[str] = None):
        self.message_id = str(uuid.uuid4())
        self.sender = sender
        self.content = content
        self.message_type = message_type
        self.recipient_id = recipient_id
        self.room_id = room_id
        self.timestamp = datetime.now()
        self.status = MessageStatus.PENDING
        self.edited_at: Optional[datetime] = None
        self.is_deleted = False
        
        # Message metadata
        self.server_id: Optional[str] = None
        self.delivery_attempts = 0
        self.max_delivery_attempts = 3
        
        # Reply information
        self.reply_to_message_id: Optional[str] = None
        self.thread_id: Optional[str] = None
        
        # Media attachments
        self.attachments: list = []
        
        # Reactions
        self.reactions: dict = {}  # user_id -> reaction_emoji
        
        # Recipient reference (for private messages)
        self.recipient: Optional[User] = None
    
    def set_recipient(self, recipient: User):
        """Set message recipient"""
        self.recipient = recipient
        self.recipient_id = recipient.user_id
    
    def mark_as_sent(self, server_id: str):
        """Mark message as sent"""
        self.status = MessageStatus.SENT
        self.server_id = server_id
    
    def mark_as_delivered(self):
        """Mark message as delivered"""
        self.status = MessageStatus.DELIVERED
    
    def mark_as_read(self):
        """Mark message as read"""
        self.status = MessageStatus.READ
    
    def mark_as_failed(self):
        """Mark message as failed"""
        self.status = MessageStatus.FAILED
    
    def increment_delivery_attempts(self):
        """Increment delivery attempts"""
        self.delivery_attempts += 1
    
    def can_retry_delivery(self) -> bool:
        """Check if message can be retried for delivery"""
        return (self.status == MessageStatus.PENDING and 
                self.delivery_attempts < self.max_delivery_attempts)
    
    def edit_content(self, new_content: str):
        """Edit message content"""
        self.content = new_content
        self.edited_at = datetime.now()
    
    def delete_message(self):
        """Delete message"""
        self.is_deleted = True
        self.content = "[Message deleted]"
    
    def add_reaction(self, user_id: str, emoji: str):
        """Add reaction to message"""
        self.reactions[user_id] = emoji
    
    def remove_reaction(self, user_id: str):
        """Remove reaction from message"""
        if user_id in self.reactions:
            del self.reactions[user_id]
    
    def get_reaction_count(self) -> int:
        """Get total reaction count"""
        return len(self.reactions)
    
    def add_attachment(self, attachment: dict):
        """Add attachment to message"""
        self.attachments.append(attachment)
    
    def is_private_message(self) -> bool:
        """Check if message is a private message"""
        return self.message_type == MessageType.PRIVATE_MESSAGE
    
    def is_room_message(self) -> bool:
        """Check if message is a room message"""
        return self.message_type == MessageType.ROOM_MESSAGE
    
    def is_system_message(self) -> bool:
        """Check if message is a system message"""
        return self.message_type == MessageType.SYSTEM_MESSAGE
    
    def get_formatted_timestamp(self) -> str:
        """Get formatted timestamp"""
        return self.timestamp.strftime("%Y-%m-%d %H:%M:%S")
    
    def get_display_content(self) -> str:
        """Get display content (truncated if too long)"""
        if self.is_deleted:
            return "[Message deleted]"
        
        max_length = 100
        if len(self.content) > max_length:
            return self.content[:max_length] + "..."
        return self.content
    
    def __str__(self):
        return f"Message(sender={self.sender.display_name}, type={self.message_type.value}, content={self.get_display_content()})"
    
    def __repr__(self):
        return self.__str__()
    
    def to_dict(self):
        """Convert message to dictionary"""
        return {
            'message_id': self.message_id,
            'sender': self.sender.to_dict(),
            'content': self.content,
            'message_type': self.message_type.value,
            'recipient_id': self.recipient_id,
            'room_id': self.room_id,
            'timestamp': self.timestamp.isoformat(),
            'status': self.status.value,
            'edited_at': self.edited_at.isoformat() if self.edited_at else None,
            'is_deleted': self.is_deleted,
            'server_id': self.server_id,
            'delivery_attempts': self.delivery_attempts,
            'reply_to_message_id': self.reply_to_message_id,
            'thread_id': self.thread_id,
            'attachments': self.attachments,
            'reactions': self.reactions,
            'recipient': self.recipient.to_dict() if self.recipient else None
        } 